### PR TITLE
Populate dependencies based on 'scope' parameter

### LINKED ISSUE #19271: [Packerina Spec] Dependency jars getting packed to balo
**Description:**
When a jar file is used as a dependency in a Ballerina module (via platform.libraries in the toml), that jar is getting packed with the balo. If the jar is proprietary (eg : SAP JCO jars), it's illegal to re-distribute like this. If we don't add this under platform.libraries in toml, the building of the module fails with CLASS_NOT_FOUND, since there're inter-op methods. 

It's better if we can have an attribute like 'scope' under platform.libraries, so that the jars are used only in the compile time and not packed with the balo.

### PR DESCRIPTION
**Pull request by suganyasuven**
Fixes #19271

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java [hunk 1]
```diff
@@ -120,6 +122,26 @@ private void copyImportedJarsForModules(BuildContext buildContext, List<BLangPac
         }
     }
 
+    private void copyPlatformLibsForModules(PackageID packageID, ExecutableJar executableJar, Path project) {
+        List<Library> libraries = manifest.getPlatform().libraries;
+        if (null == libraries) {
```
Comment (praveennadarajah) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:127]:
```suggestion
        if (libraries == null) {
```

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java [hunk 2]
```diff
@@ -120,6 +122,26 @@ private void copyImportedJarsForModules(BuildContext buildContext, List<BLangPac
...
+            return;
+        }
+        List<Library> libs = libraries.stream().filter(lib -> lib.getModules() == null ||
+                Arrays.asList(lib.getModules()).contains(packageID.name.value)).collect(Collectors.toList());
+        for (Library lib : libs) {
+            String libFilePath = lib.getPath();
+            if (null != libFilePath) {
```
Comment (praveennadarajah) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:134]:
```suggestion
            if (libFilePath != null) {
```

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    public void setup() throws IOException {
+        super.setup();
+        try {
+            this.testResources = super.tmpDir.resolve("scope-test-resources");
+            testResourcesURI = getClass().getClassLoader().getResource("test-resources").toURI();
+            Path storedJarDependencyProject = Paths.get(testResourcesURI).resolve("stored-jar-dependency-project");
+            Files.walkFileTree(storedJarDependencyProject, new BuildCommandTest.Copy(storedJarDependencyProject,
+                    this.testResources));
+        } catch (URISyntaxException e) {
+            Assert.fail("error loading resources");
```
Comment (praveennadarajah) [inline cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:62]:
```suggestion
            Assert.fail("Error loading resources");
```

Comment (suganyasuven) [reply cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:62]: Error messages start with simple letters.

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        Assert.assertFalse(isJarExists(baloZipFile, storedJarPath));
+    }
+
+    private boolean renameFile(File oldName, File newName) {
+        return oldName.renameTo(newName);
+    }
+
+    private boolean isJarExists (File file, String jarEntry) throws IOException {
+        try (ZipFile zipFile = new ZipFile(file)) {
+            Enumeration<? extends ZipEntry> entries = zipFile.entries();
```
Comment (praveennadarajah) [inline cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:128]: No need to iterate all the zip entries and find whether a particular entry is there or not. The time complexity and performance hit would be high by doing this way. Instead, you can directly check a particular entry's presence in the following way.

`return new ZipFile(file).getEntry(jarEntry) != null;`

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        copy(scopeToml.resolve("case2").resolve(ProjectDirConstants.MANIFEST_FILE_NAME),
+                this.testResources.resolve(ProjectDirConstants.MANIFEST_FILE_NAME));
+        buildCommand.execute();
+        Assert.assertTrue(Files.exists(balo), "Check if balo directory exists");
+
+        // Check whether dependency jars getting packed to balo
+        Assert.assertTrue(renameFile(baloFile, baloZipFile));
+        Assert.assertFalse(isJarExists(baloZipFile, storedJarPath));
+
+        // compile scope
```
Comment (IrushiL) [inline cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:109]: When checking the Ballerina.toml file, it looks like "case3" should be the "test" scope.

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        copy(scopeToml.resolve("case1").resolve(ProjectDirConstants.MANIFEST_FILE_NAME),
+                this.testResources.resolve(ProjectDirConstants.MANIFEST_FILE_NAME));
+        buildCommand.execute();
+        Assert.assertTrue(Files.exists(balo), "Check if balo directory exists");
+
+        // Check whether dependency jars getting packed to balo
+        Assert.assertTrue(renameFile(baloFile, baloZipFile));
+        Assert.assertTrue(isJarExists(baloZipFile, storedJarPath));
+
+        // package scope
```
Comment (IrushiL) [inline cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:99]: When checking the Ballerina.toml file, it looks like "case2" should be the "compile" scope.

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        copy(scopeToml.resolve("case1").resolve(ProjectDirConstants.MANIFEST_FILE_NAME),
+                this.testResources.resolve(ProjectDirConstants.MANIFEST_FILE_NAME));
+        buildCommand.execute();
+        Assert.assertTrue(Files.exists(balo), "Check if balo directory exists");
+
+        // Check whether dependency jars getting packed to balo
+        Assert.assertTrue(renameFile(baloFile, baloZipFile));
+        Assert.assertTrue(isJarExists(baloZipFile, storedJarPath));
```
Comment (IrushiL) [inline cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:97]: This code segment seems to be repeating for all 3 scopes. Is it possible to separate it out and avoid code duplication?

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/toml/model/Library.java [hunk 8]
```diff
@@ -68,4 +69,12 @@ public String getArtifactId() {
     public void setArtifactId(String artifactId) {
         this.artifactId = artifactId;
     }
+
+    public String getScope() {
+        return scope;
+    }
+
+    public void setScope(String scope) {
```
Comment (IrushiL) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/toml/model/Library.java:77]: This method does not seem to be used.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/toml/model/Library.java [hunk 9]
```diff
@@ -28,6 +28,7 @@
     public String path;
     public String groupId;
     public String[] modules;
+    public String scope;
```
Comment (IrushiL) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/toml/model/Library.java:31]: Can we make this a private field?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java [hunk 10]
```diff
@@ -120,6 +122,26 @@ private void copyImportedJarsForModules(BuildContext buildContext, List<BLangPac
...
+        if (libraries == null) {
+            return;
+        }
+        List<Library> libs = libraries.stream().filter(lib -> lib.getModules() == null ||
+                Arrays.asList(lib.getModules()).contains(packageID.name.value)).collect(Collectors.toList());
+        for (Library lib : libs) {
+            String libFilePath = lib.getPath();
+            if (libFilePath != null) {
+                Path nativeFile = project.resolve(Paths.get(libFilePath));
+                if (lib.getScope() != null && lib.getScope().equals("test")) {
```
Comment (warunalakshitha) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:136]: if lib.getScope() == null means, default version? based on the logic scope null values are put in to moduleLibs.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/writer/BaloFileWriter.java [hunk 11]
```diff
@@ -298,15 +295,15 @@ private void addModuleDoc(Path root, Path moduleSourceDir) throws IOException {
...
-                                 HashSet<Path> moduleDependencyList) throws IOException {
+    private void addPlatformLibs(Path root, Path projectDirectory, String moduleName) throws IOException {
         //If platform libs are defined add them to balo
         if (null != manifest.getPlatform().libraries) {
             Path platformLibsDir = root.resolve(ProjectDirConstants.BALO_PLATFORM_LIB_DIR_NAME);
             Files.createDirectory(platformLibsDir);
 
             List<Path> libs = manifest.getPlatform().libraries.stream()
-                    .filter(lib -> lib.getModules() == null || Arrays.asList(lib.getModules()).contains(moduleName))
+                    .filter(lib -> (lib.getModules() == null || Arrays.asList(lib.getModules()).contains(moduleName))
```
Comment (warunalakshitha) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/writer/BaloFileWriter.java:305]: Shall we replace with for loop without using java streams.

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java (hunk 7 above)
Comment (warunalakshitha) [reply cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:97]: May be we can have three test cases for each scope which will be more clear

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java (hunk 10 above)
Comment (suganyasuven) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:136]: yes. We don't need that non-null check here.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/toml/model/Library.java (hunk 8 above)
Comment (suganyasuven) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/toml/model/Library.java:77]: Yes we can remove the setter

File: cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+import java.nio.file.Paths;
+import java.util.zip.ZipFile;
+
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.wso2.ballerinalang.compiler.util.ProjectDirConstants.BLANG_COMPILED_PKG_BINARY_EXT;
+
+/**
+ * Dependency scope tests.
+ *
+ * @since 1.2.x
```
Comment (warunalakshitha) [inline cli/ballerina-packerina/src/test/java/org/ballerinalang/packerina/cmd/DependencyScopeTest.java:43]: Should be 1.3.0

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java [hunk 13]
```diff
@@ -239,6 +243,42 @@ private void copyLibsFromBalo(Path baloFilePath, Set<Path> moduleDependencySet)
         }
     }
 
+    private void copyAndValidateBaloDependencies(BuildContext buildContext, Path importDependencyPath,
```
Comment (warunalakshitha) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:246]: If I import 2 modules (mod1, mod2) and mod1 has compiler dependency and same compiler dependency add as default in mod2. But that dependency will be packed in mod2 balo and at runtime everything will work fine. But still warning message will be shown during compilation as per below logic. Shall we fix that case as well

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java [hunk 14]
```diff
@@ -239,6 +243,42 @@ private void copyLibsFromBalo(Path baloFilePath, Set<Path> moduleDependencySet)
...
+        Manifest manifestFromBalo = RepoUtils.getManifestFromBalo(importDependencyPath);
+        List<Library> baloDependencies = manifestFromBalo.getPlatform().libraries;
+
+        if (baloDependencies == null) {
+            return;
+        }
+
+        // If platform libs are defined, copy them to target
+        List<Library> libraries = manifest.getPlatform().libraries;
+        for (Library baloTomlLib : baloDependencies) {
```
Comment (warunalakshitha) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:257]: This logic has performace impact due to loop inside loop. Can we collect all compiler libs of balos in ahash set( key can be orgname+ packageid+ artfactID+ filename) and default libs two another set. Then we can compare those two hashsets two check first one is subset of default(packing) libs.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java (hunk 13 above)
Comment (suganyasuven) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:246]: Still, we need that warning message as dependencies of mod2 will be copied to the  mod2 and the current module only. For compiling mod1, we need to copy that dependency to mod1. As we are copying the dependencies based on the modules, the user have to define that dependency for mod1. Please correct me if I'm wrong.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java [hunk 15]
```diff
@@ -239,6 +244,46 @@ private void copyLibsFromBalo(Path baloFilePath, Set<Path> moduleDependencySet)
...
+            }
+        }
+
+        if (platformLibs.containsAll(baloCompileScopeDependencies)) {
+            return;
+        }
+
+        baloCompileScopeDependencies.removeAll(platformLibs);
+        for (Path baloTomlLib : baloCompileScopeDependencies) {
+            buildContext.out().println("warning: " + packageID + " is missing a native library dependency - " +
```
Comment (warunalakshitha) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CopyNativeLibTask.java:282]: can we check !platformLibs.contains(baloTomlLib) and then log warning. so we can get rid of 276 to 280 lines

**Pull request merged**